tests/test_dunder.rs

```diff
@@ -552,3 +553,77 @@ fn getattr_doesnt_override_member() {
     py_assert!(py, inst, "inst.data == 4");
     py_assert!(py, inst, "inst.a == 8");
 }
+
+/// Wraps a Python future and yield it once.
+#[pyclass]
+struct OnceFuture {
+    future: PyObject,
+    polled: bool,
+}
+
+#[pymethods]
+impl OnceFuture {
+    #[new]
+    fn new(future: PyObject) -> Self {
+        OnceFuture {
+            future,
+            polled: false,
+        }
+    }
+}
+
+#[pyproto]
+impl PyAsyncProtocol for OnceFuture {
+    fn __await__(slf: PyRef<Self>) -> PyResult<Py<Self>> {
+        Ok(slf.into())
+    }
+}
+
+#[pyproto]
+impl PyIterProtocol for OnceFuture {
+    fn __iter__(slf: PyRef<Self>) -> PyResult<Py<Self>> {
+        Ok(slf.into())
+    }
+    fn __next__(mut slf: PyRefMut<Self>) -> PyResult<Option<PyObject>> {
```

Comment:
Just curious: can this be `fn __next__(&mut self) -> Option<PyObject>`  ?

Reply:
We can't do so since `None` is converted to `exceptions::StopIteration`.
Python C-API allow `yield` some value by wrapping it by `StopIteration` so maybe it's better to return `PyResult<Either<Future, Result>>`. But it would too much for this PR.

Reply:
Ah, ok sounds good :+1: